doc/user-guide/testing.rst

```diff
@@ -0,0 +1,178 @@
+that you did not even think to look for!
+
+Strategies
+~~~~~~~~~~
+
+Each source of examples is called a "strategy", and xarray provides a range of custom strategies which produce xarray
+data structures containing arbitrary data. You can use these to efficiently test downstream code,
+quickly ensuring that your code can handle xarray objects of all possible structures and contents.
+
+These strategies are accessible in the :py:module::`xarray.testing.strategies` module, which provides
+
+.. currentmodule:: xarray
+
+.. autosummary::
+
+   testing.strategies.valid_dtypes
+   testing.strategies.np_arrays
+   testing.strategies.names
+   testing.strategies.dimension_names
+   testing.strategies.dimension_sizes
+   testing.strategies.variables
+   testing.strategies.coordinate_variables
+   testing.strategies.dataarrays
+   testing.strategies.data_variables
+   testing.strategies.datasets
+   testing.strategies.chunks
+   testing.strategies.chunksizes
+
+Generating Examples
+~~~~~~~~~~~~~~~~~~~
+
+To see an example of what each of these strategies might produce, you can call one followed by the ``.example()`` method,
+which is a general hypothesis method valid for all strategies.
+
+(TODO we should specify a seed to hypothesis so that the docs generate the same examples on every build)
+
+.. ipython:: python
+
+    import xarray.testing.strategies as xrst
+
+    xrst.dataarrays().example()
+    xrst.dataarrays().example()
+    xrst.dataarrays().example()
+
+You can see that calling ``.example()`` multiple times will generate different examples, giving you an idea of the wide
+range of data that the xarray strategies can generate.
+
+In your tests however you should not use ``.example()`` - instead you should parameterize your tests with the
+``hypothesis.given`` decorator:
+
+# TODO finishsimple test example
+
+.. ipython:: python
+
+    from hypothesis import given
+
+.. ipython:: python
+
+    @given(xrst.dataarrays())
+    def test_something(da):
+        ...
+
+
+Chaining Strategies
+~~~~~~~~~~~~~~~~~~~
+
+Xarray's strategies can accept other strategies as arguments, allowing you to customise the contents of the generated
+examples.
+
+.. ipython:: python
+    :okexcept:
+
+    # generate a DataArray with shape (3, 4), but all other details still arbitrary
+    xrst.dataarrays(
+        data=xrst.np_arrays(shape=(3, 4), dtype=np.dtype("int32"))
+    ).example()
+
+This also works with strategies defined in other packages, for example the ``chunks`` strategy defined in
+``dask.array.strategies``.
+
+.. warning::
+    Passing multiple different strategies to the same constructor can lead to poor example generation performance.
+
+    This is because in order to construct a valid xarray object to return, our strategies must check that the
+    variables / dimensions / coordinates are mutually compatible. We do this using ``hypothesis.assume``, which throws
+    away any generated examples not meeting the required condition.
```

Comment:
> Unfortunately this will perform catastrophically badly ("the API is useful on few realistic problems"), so I'd strongly recommend finding an alternative.

Yeah this was the main thing I was worried about with this PR.

> question: how do dimensions and coordinates relate in a DataSet?

It's better to think of it the way it really works: `Datasets` don't actually contain `DataArrays`, instead they contain many `Variable` objects, some of which are designated to be "Coordinate Variables". All `Variables` in the `Dataset` must depend on (some subset of) the same shared set of dimensions (with consistent lengths). E.g. if `dataset.sizes = {"x": 2, "y": 3}`, then all data/coordinate variables in that dataset must have `sizes = {"x": 2, "y": 3}`, `{"x": 2}`, `{"y": 3}`, or `{}`. When you pull out a DataArray via `ds["name"]` really what happens is that a new DataArray is built from the variable you selected, and all the other coordinates in the dataset get attached on with it to form the returned DataArray.

> need not have them in the same order

Order is not relevant once the xarray object has been created. Order *does* matter during initialisation, where you need to make sure that the shape of the numpy array and the dims match up.

Coordinate variables are basically the same thing as data variables, with one exception. If a coordinate is 1D, and is named the same as one of the dimensions, it is a "dimension coordinate" (bad name yes). This is a coordinate, but it also implies that a `pandas.Index` object has been created behind the scenes, which allows for fast `.sel()` lookups.

(This would all be much clearer if we had actually properly [publicly exposed the lower-level `Variable` class](https://docs.xarray.dev/en/stable/roadmap.html#labeled-array-without-coordinates), and once the [Flexible Indexes refactor](https://docs.xarray.dev/en/stable/roadmap.html#flexible-indexes) stuff fully exposes the underlying `pd.Index` objects.)

> Is it possible to also have them offset along a shared dimension with coordinate, e.g. different arrays for each hemisphere?

Not sure exactly what you mean, but the fact that within a dataset a dimension named `"x"` must always have the same length is clearly restrictive, and I wrote [xarray-datatree](https://github.com/xarray-contrib/datatree) to move past that restriction.

Reply:
> I suspect we're in for some ongoing discussion here - apologies for the bad news, but I'm up for it if you are.

I'm up for it!

> Accept a mapping of kwargs to pass through to 'inner' strategies. Used all over matplotlib (albeit as **kwargs), and it's an option for the elements= argument to npst.arrays().

I don't think this would actually solve the performance problem in this case would it? We would still have to `assume()` to check that the given kwargs were compatible.

> - Accept the union of all the parameters for the inner strategies, raise an error if they're incompatible, and derive a consistent+performant strategy. Nice in principle, often intractable in practice to support everything users want.

I'm struggling with all the possible cases when only accepting `data`, `dims`, and `coords`, so I don't think we can feasibly expand this to all other possible parameters too.

> Just provide primitives, and let users work out how to use them.

I mean I definitely want to provide `datasets()`. For more specific use cases I would be okay just making users reconstruct their own strategies though. In fact I'm actually leaning towards this option.

> Accept a [strategy for] a sequence of dimensions, and strategies for individual coordinates and variables; then generate exactly the right number of the latter internally. This is probably the cleanest and most-performant solution if the UX is acceptable, but it's pretty abstract.

This could work, but what is "the right number"?